Commands/BuildDeps.cs

```diff
@@ -90,12 +101,115 @@ protected override int Execute()
+                lock (buildedDepsNames)
+                {
+                    if (NoNeedToBuild(module, modulesToBuild))
+                    {
+                        builtStorage.AddBuiltModule(module, modulesOrder.CurrentCommitHashes);
+                        buildedDepsNames.Add(module.Name);
+                        continue;
+                    }
+                    if (isFailed) break;
+                }
+
+                semaphore.Wait();
+                Log.Debug($"Waiting for {module.ToBuildString()}");
+                
+                while (true)
+                {
+                    lock (buildedDepsNames)
+                    {
+                        
+                        if (isFailed) break;
+                        // Может быть так, что мы уже собрали module/client и сейчас начинаем собирать module/full-build, т.е. пересборка по сути.
+                        // Сборка развалится если в этот же момент начнут собираться те, кто зависят от module/client.
+                        // Для пересобираемых модулей ждем когда все остальные таски отработают для гарантии того, что не сломаем текущие сборки.
+                        // И гарантируем, что не начнет собираться другой модуль, который зависит от пересобираемого.
+                        // В общем случае такие проблемы могут возникнуть с прямыми зависимостями модуля.
+                        // Но некоторе модули не указывают все свои прямые зависимости. Поэтому приходится смотреть всё дерево зависимостей.
+                        var allDepsIsBuilded = AllModuleDepsIsBuilded(module, buildedDepsNames, modulesOrder);
+                        var isRebuildAlreadyBuildedModule = buildedDepsNames.Contains(module.Name);
+                        
+                        if (allDepsIsBuilded && isRebuildAlreadyBuildedModule && semaphore.CurrentCount == Helper.MaxDegreeOfParallelism - 1)
+                        {
+                            buildedDepsNames.Remove(module.Name);
+                            break;
+                        }
+                        if (allDepsIsBuilded && !isRebuildAlreadyBuildedModule) break;
+
+                    }
+                    Task.WaitAny(tasks.Where(t => !t.IsCompleted).ToArray());
+                }
+
+                if (isFailed)
+                {
+                    semaphore.Release();
+                    break;
+                }
+
+                tasks.Add(Task.Run(() =>
+                {
+                    ConsoleWriter.WriteProgress($"{module.ToBuildString(),-49} {$"{built}/{modulesToBuild.Count}",10}");
+                    
+                    try
+                    {
+                        Log.Debug($"Building for {module.ToBuildString()}");
+                        if (!builder.Build(module)) isFailed = true;
+                    }
+                    catch (Exception e)
+                    {
+                        exception = e;
+                        isFailed = true;

```

Comment:
Why not just throw and remove `isFailed` flag?

Reply:
Because we want to stop the build process as early as possible. Otherwise, all other modules will try to build despite an error.